Replace Ruby with Kotlin

garyttierney opened the pull request:
As discussed in #310 this pull request introduces a new Kotlin based plugin system as well as a much, much better plugin build system.

Major planned features:

- [x] Debugging support (with IntelliJ or Eclipse's step-through debugger for example)
- [x] Unit test support
- [x] Integration test framework
- [x] Full editor integration (auto-completion, error reporting, etc.)

For a more detailed explanation of what we get with kotlin and this new plugin system see the discussion in #310.

As well as the plugin system itself, there is a lot of plugins written in Ruby left to port.  A list of these is below.

- [x] areas
- [x] bank
- chat
  - [x] privacy
  - [x] private-messaging
- cmd
  - [x] animate
  - [x] bank
  - [x] item
  - [x] lookup
  - [x] messaging
  - [x] npc
  - [x] punishment
  - [x] skill
  - [x] teleport
- [x] consumables
- [x] dummy
- [x] emote-tab
- entity
  - [x] attributes
  - mob
    - [x] extension
    - [x] following
    - [x] walk-to
  - [x] spawning
- location
  - [x] al-kharid
  - [x] edgeville
  - [x] falador
  - [x] lumbridge
  - [x] varrock
- [x] logout
- navigation
  - [x] door
- [x] player-action
- [x] quest
- [x] run
- [x] shops
- skill
  - [x] fishing
  - [x] herblore
  - [x] mining
  - [x] prayer
  - [x] runecraft
- [x] util (now part of the api plugin)

Two pugins have been deselected from the kotlin port, and will be done at a later date - dialogue and the (non-combat) magic skill.

This doesn't include tests for the plugins - see #361.

garyttierney committed:
First draft of KotlinPluginEnvironment

garyttierney committed:
Add a 'plugins' source set to the game module

Creates a new sourceset for server plugins to support IDE integration
with plugin code.  Additionally moves plugins to game/data/plugins and
contains some gradle related fixes (remove java plugin from root project
and fix exec task dependency).

garyttierney committed:
Compile plugins at build-time instead of runtime

Adds gradle tasks to build all plugin scripts under data/plugins with
the KotlinPluginCompiler implementation previously used for runtime code
generation.  In addition to .plugin.kts files, scripts can also declare
API code in .kt files which will also be included on the classpath and
made available to other plugins.

garyttierney committed:
Reorganize project sources from src/main to src/main/java

garyttierney committed:
Remove 'world' object from global script scope

garyttierney committed:
Add port of the 'dummys' plugin from Ruby

garyttierney committed:
Add private messaging plugin

garyttierney committed:
Treat each plugin as an individual source set

Adds separate build tasks for each plugin by auto-discovering plugin meta
files in the build script.  Each plugin will automatically have its
main sources and tests compiled, and then it's output added to the game
modules classpath.

This enables support for incremental compilation of scripts, as well as
unit testing using Gradle's test framework.

garyttierney committed:
Add plugin tests to the game build

ryleykimmel commented:
During this conversion I think we should go through and fix any open plugin issues left over from Ruby (such as the bugs that have been found, etc)

Promises commented:
Could also add woodcutting, since it's just mining with a few different ids, I could do it myself as soon as mining is done.

garyttierney committed:
Add stub to plugin compile-time classpath

garyttierney commented:
@ryleykimmel ACK

@Promises I'd rather we didn't add new work onto this already monolithic task.  However, am happy to merge code into it as we go if people would like to take on plugin work themselves.

Major- commented:
@Promises feel free to work on wc yourself and we'll offer code review + happily accept a pull request

@garyttierney `util` might not need to exist, need to think about what stuff to put in the (kotlin version of) bootstrap

garyttierney commented:
@Major- right.  Probably some core things we want to create bindings for in the bootstrap.  Some thoughts:

* Scheduled tasks - currently only used for following - per-player scheduled task or action might work better
* Message Listeners - already done
* Event Listeners - todo
* Attribute bindings - todo - I think the bootstrap should provide this since attributes are very "core" (we need to persist them).
* Game entity lookups

garyttierney committed:
Move stub.kt to main game module

WizardJesse1 committed:
Add height and direction params to npc_spawn

WizardJesse1 committed:
Port location NPC spawns from Ruby to Kotlin

cubeee commented:
I'll help convert some plugins today

garyttierney committed:
Make run task depend on plugin scripts

garyttierney committed:
Clean up plugin script compilation

garyttierney committed:
Fix builds on other platforms (non-Unix)

cubeee commented:
Seems that plugins with identical file names can't be used even if they are in other packages. 
I had a file called bank.plugin.kts for banking commands and it wouldn't load because of bank plugin's identically named bank.plugin.kts file

cubeee committed:
Add base for Kotlin commands

cubeee committed:
Simplify on_command arguments

cubeee committed:
Convert ::animate command to Kotlin

cubeee committed:
Convert ::item command to Kotlin

cubeee committed:
Convert messaging commands to Kotlin

cubeee committed:
Overload valid_arg_length

cubeee committed:
Convert punishment commands to Kotlin

cubeee committed:
Convert teleport commands to Kotlin

cubeee committed:
Convert experience commands to Kotlin

cubeee committed:
Convert npc spawn commands to Kotlin

Major- commented:
@garyttierney ^

Jire commented:
Cx

cubeee committed:
Convert lookup commands to Kotlin

cubeee committed:
Add input type checking to commands

garyttierney committed:
Merge pull request #340 from cubeee/kotlin-port/commands

Convert commands to Kotlin

garyttierney committed:
Refactor the plugin compile task for performance

Integrates the plugin script compilation task with Gradle so we no
longer need to start a new instance of the JVM for each set of plugin
scripts that we want to compile.

garyttierney committed:
Make plugin classes dependant on scripts

garyttierney committed:
Check for conflicting plugin names at build-time

garyttierney committed:
Update travis-ci configuration to cache build deps

garyttierney committed:
Add missing name normalization to lookup plugin

garyttierney committed:
Add start of test framework for plugins

Adds a basic testing framework suitable for plugins that start simple
Actions for players, which can be expanded on in the future.  The
banking and training dummy tests have been updated to use this framework
and serve as samples.

garyttierney committed:
Remove test context from plugin testing framework

garyttierney commented:
Lots of work done on testing now.  Any thoughts so far?  @Major- @ryleykimmel

garyttierney committed:
Allow multiple players per test

Refactors the test helpers to use receiver functions so multiple players
can exist in the world per test case.  The AddFriendsTest is an example
of where this is needed.

Additionally, Hamcrest has been replaced with AssertJ for fluent
assertions.

garyttierney committed:
Add support for asynchronous Mob actions

Adds asynchronous implementations of the Action and DistancedAction
classes, allowing plugins to make use of suspendable functions in Kotlin
instead of creating mini state machines to keep track of state.

The asynchronicity works by creating coroutine backed Channels for each
asynchronous action and having them listen on "pulse" events from the
action scheduler.  The action can then suspend execution until a pulse is
received or until some expensive operation has completed (i.e.,
pathfinding).

If an asynchronous action is still busy when a pulse arrives then the
number of missed pulses will be accumulated and sent to the action at
the next possible time.

The training dummy plugin has been updated to use asycnrhonous actions
as an example.

garyttierney committed:
Add Jenkinsfile

garyttierney committed:
Add plugin for consumable food/drink

garyttierney committed:
Port the walkto and following plugins to Kotlin

garyttierney committed:
Add on_button method to plugin scripts

garyttierney committed:
Add Kotlin port of the logout plugin

garyttierney committed:
Add kotlin port of the emote-tab plugin

garyttierney committed:
Add kotlin port of the walk/run toggle plugin

tlf30 committed:
Adding support for mining skill

tlf30 committed:
Finish up porting mining plugin to Kotlin

tlf30 commented:
I have created a merge request for a port of the mining plugin over to Kotlin. 
If it is all good and approved I will work on porting over more plugins. 

Also @Promises, if this goes through, could you possibly use it for wood cutting?

Promises commented:
@tlf30  its pretty much the same as mining, so i'll take a look.

tlf30 commented:
@Promises I had some free time so I build a woodcutting plugin. I do not have all of the ID perfect yet. I was wondering if you could post on the PR some ID numbers for trees/stumps if you had them. 
Thanks,
Trevor

Commit:
Share kotlin version config between buildSrc and project (#353)

* Share kotlin version config between buildSrc and project
* Use built-in Kotlin MessageCollector

garyttierney committed:
Fix plugin test runner

garyttierney committed:
Remove plugins.gradle and factor into common extension

garyttierney committed:
Fix plugin test running with new gradle modules

garyttierney committed:
Fix incremental script compilation

garyttierney committed:
Use main sourcset output dir for scripts

garyttierney committed:
Use File.separator in settings.gradle

garyttierney committed:
Add an await function to the AsyncActionRunner trait

Major- committed:
Fix plugin processing on windows

garyttierney committed:
Add API plugin with common functionality

garyttierney committed:
Refactor mining to use async actions

Major- committed:
Add kotlin to .editorconfig

Uses the recommended conventions.

garyttierney committed:
Add explicit stop() to async actions

Commit:
Merge pull request #354 from apollo-rsps/editorconfig

Add kotlin to .editorconfig

garyttierney committed:
Add basic detekt configuration

tlf30 committed:
Initial commit on new woodcutting branch

tlf30 committed:
Woodcutting now uses Async Action

tlf30 committed:
Fix all issues. Wood cutting now works.

garyttierney committed:
Rewrite package names in compiled script files

garyttierney committed:
Add source information to remapped scripts

garyttierney committed:
Enable experimental coroutines for all kotlin projects

garyttierney committed:
Clear static object REMOVE messages when re-adding

garyttierney committed:
Refactor asynchronous actions to avoid thread usage

Adds custom coroutines specifically for use within Actions.  This
implements all features of the previous system, however, does not rely
on a separate executor to wake up continuations.

Note: since Actions are still ran by an executor and a small chance of
overlap exists the ActionCoroutine implementation uses atomic counters
and references.

Major- commented:
With kotlins extension properties I don't think we need the attributes plugin anymore, although we do need to consider how we serialize extension data.

garyttierney committed:
Refactor woodcutting to use coroutines

Commit:
Merge pull request #355 from tlf30/kotlin-experiments-woodcutting

Added woodcutting plugin for Kotlin

arinhouck committed:
Add door support in Kotlin (#356)

Adds support for doors as a Kotlin plugin, based on the previous ruby plugin by shiver.

Major- committed:
Support lookup_x inputs with the id at the end

Major- committed:
Add shops plugin

Thanks to tlf30 for a lot of work on this.

tlf30 committed:
Add varrock shops

garyttierney committed:
Update varrock plugin package name

Commit:
Merge pull request #360 from apollo-rsps/kotlin-shops

Add shops in kotlin

Major- committed:
Rename dls.kt to dsl.kt

tlf30 committed:
Port Fishing plugin to kotlin (#347)

This also fixes a bunch of bugs.

Major- committed:
Remove meta.toml files

Since plugins were moved to gradle they aren't used anyway.

Major- committed:
Add destructuring operators for Position

Major- committed:
Let plugins intercept Events

Support for PlayerEvents already exists, but general event
support wasn't included in the original KotlinPluginScript.

Major- committed:
Port Area plugin to Kotlin

Major- committed:
Add skill extension properties

Major- committed:
Add packages to plugins

Major- committed:
Correctly import api.Definitions

Major- committed:
Merge pull request #366 from apollo-rsps/kotlin-experiments-areas

Port the area actions plugin

Commit:
Add Slack invite link to the README

tlf30 committed:
Add implementation of runecrafting skill

tlf30 committed:
Ported prayer plugin to kotlin

Major- committed:
Fix nitpicks in Fishing plugin

Major- committed:
Remove incorrect Gunnarsgrunn fishing spots

The spots are actually on Tutorial Island and result in two
fishing spots stacked on top of each other.

garyttierney committed:
Remove dependency on embedded kotlin compiler

This is a large change that significantly simplifies the plugin build system by
having the kotlinc handle everything related to script compilation.  Since we
no longer use package name rewriting, the "packageName" option has been dropped
from the gradle script API and has been removed from existing plugins.

In addition to simplifying the build system, it should also significantly speed
up build times.  Previously we ran 2 build processes for every plugin and
didn't have fully working incremental compilation for scripts compared to 1
build process with built-in incremental compilation.

There were also some random pieces of code that referenced kotlinc directly, so
those have either been removed or refactored out.

garyttierney committed:
Add IntelliJ kotlinc configuration

This removes the need for the plugin stub.

garyttierney committed:
Remove plugin script stub

Major- committed:
Fix nitpicks in Woodcutting plugin

Major- committed:
Fix nitpicks in Mining plugin

Also adds equals() and hashcode() implementations to the various
Mining actions.

Major- committed:
Add prospecting on expired rocks

Major- committed:
Separate prospecting actions

Major- committed:
Improve ::position command message

Major- committed:
Merge pull request #386 from Major-/kotlin-experiments

Improve woodcutting, fishing, mining plugin code

Major- committed:
Merge pull request #387 from Major-/kotlin-experiments-pos

Improve ::position command message

Major- committed:
Move Position destructuring functions to their own object

tlf30 committed:
Fix to allow multiple prayers.
Fixed not caching enum values.

garyttierney committed:
Remove now invalid packageName from prayer plugin

garyttierney committed:
Fix code-style in prayer plugin

Adds named arguments to enum declarations along with lookup functions, fixes
continuation indent, and replaces Map<Player, ArrayList<Prayer>> with a
SetMultimap<Player, Prayer>.

garyttierney committed:
Clean up remaining issues on prayer plugin

garyttierney committed:
Use a map for bone and prayer lookups

Major- committed:
Rename npc spawning scripts

Major- committed:
Optimise Mob.walkTo for paths that require no verification

Major- committed:
Fix nitpicks in emote plugin

cfoust committed:
Add teleportation to places using ::tele [name]

Major- committed:
Fix scriptTemplatesClasspath for most users

Users will have to build using IntelliJ at least once, but this
should reflect the workflow of the majority of users.

Major- committed:
Merge pull request #391 from cfoust/feat/teleport-shortcuts

Chivvon committed:
Port Herblore plugin to Kotlin

Major- committed:
Merge pull request #396 from chivvon/herblore

Major- committed:
Fix nitpicks in prayer plugin

Major- committed:
Add kotlin continuation indent to editorconfig

Major- committed:
Remove command utilities plugin

Major- committed:
Fix typo in shops dsl

Major- committed:
Remove util:lookup plugin

Behaviour moved into the api plugin.

Major- committed:
Add gradle to .editorconfig

stevesoltys committed:
Fix numerical attribute encoding

stevesoltys committed:
Optimize collision matrix building

stevesoltys committed:
Add support for checking the bounds of entities

Arham4 committed:
Added all shops in Lumbridge, Falador, Edgeville, and Al-Kharid

stevesoltys committed:
Fix A* pathfinding when height is greater than zero

garyttierney committed:
Update to Kotlin 1.60 and Gradle 4.9

Removes the dependency on legacy script compilation and now relies on script
discovery.  In addition, the Gradle build scripts were refactored and updated
to be compatible with Gradle 5.0 and make use of the new java-library
configurations.

garyttierney committed:
Add core plugins to the servers runtime classpath

garyttierney committed:
Make buildscripts backwards compatible with Gradle 4.0

garyttierney committed:
Remove Detekt configuration

garyttierney committed:
Remove Kotlin scripting Gradle plugin

garyttierney committed:
Replace fast-classpath-scanner with classgraph

garyttierney committed:
Update plugin test framework to junit5

Updates the testing infrastructure to use the latest relesae of junit and
leverages the new extension mechanism to create an easy to use testing
framework.  Also adds additional test coverage for several plugins.

garyttierney committed:
Use Gradles built-in JUnit 5 support

garyttierney committed:
Update test runner and coverage for CI

garyttierney committed:
Add missing wrapper task in Travis config

garyttierney committed:
Put plugins with the same name in unique packages

garyttierney committed:
Produce coverage report with JaCoCo during CI

garyttierney committed:
Fix recording of code coverage

garyttierney committed:
Produce XML coverage report for codecov.io

Major- committed:
Throw custom exception on missing plugin dependency

Major- committed:
Remove (empty) util plugin

The contents of this plugin have mostly been moved into the api plugin,
or removed entirely.

Major- committed:
Rename npc_spawn function to spawnNpc

Major- committed:
Add unit tests for areas plugin

Major- committed:
Add tests for player actions

Major- committed:
Support npc and object definitions in plugin tests

tlf30 commented:
What is left for work to get this merged?

Major- commented:
@tlf30 At long last, not much! We want comprehensive plugin tests (#361), ways to make the testing plugins more ergonomical (#382, #395), and documentation for developing plugins (#393).

garyttierney committed:
Add Gradle wrapper

garyttierney committed:
Add Appyveyor configuration

garyttierney committed:
Create zip file distributions for each AppVeyor build

garyttierney committed:
Remove Gradle wrapper hack from Travis config

garyttierney committed:
Add WalkingQueue tests

garyttierney committed:
Include default configuratio in app distribution

garyttierney committed:
Add ktlint Gradle plugin

garyttierney committed:
Remove empty allprojects block

garyttierney committed:
Fix inclusion of ktlint for check tasks

Major- committed:
Merge pull request #407 from arham4/kotlin-experiments

garyttierney committed:
Merge remote-tracking branch 'origin/master' into kotlin-experiments

Major- committed:
Repackage api plugin

garyttierney committed:
Add SonarCloud scans to Travis runs

garyttierney committed:
Fix order of gradle arguments in sonarqube run

garyttierney committed:
Run Sonarqube after generating JaCoCo report

garyttierney committed:
Use travis built-in SonarCloud support

garyttierney committed:
Move shop plugins to unique packages

Major- committed:
Make plugins target java 8

Major- committed:
Add @DefinitionSource annotations

Allows parameterized tests to use Item, Npc, and Object definitions
as @ArgumentSources.

This commit also adds support for using @ItemDefinitions etc on
properties as well as functions.

Major- committed:
Add tests for api plugin

Major- committed:
Add tests for spawn plugin

Major- committed:
Add tests for fishing plugin

garyttierney commented on game/plugin/skills/fishing/src/org/apollo/game/plugin/skills/fishing/FishingAction.kt:
```diff
@@ -0,0 +1,88 @@
+     */
+    private val tool = target.option.tool
+
+    override fun action(): ActionBlock = {
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. ([kotlin:S3776](https://sonarcloud.io/organizations/github-apollo-rsps/rules#rule_key=kotlin%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=apollo%3Aorg.apollo&pullRequest=338&issues=AWVzgXw2XKENQn1Ssjbk&open=AWVzgXw2XKENQn1Ssjbk)

[ISSUE_KEY]: AWVzgXw2XKENQn1Ssjbk

garyttierney commented on game/plugin/skills/fishing/src/org/apollo/game/plugin/skills/fishing/FishingTarget.kt:
```diff
@@ -0,0 +1,39 @@
+
+    /**
+     * Returns whether or not the catch was successful.
+     * TODO: We need to identify the correct algorithm for this
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Complete the task associated to this TODO comment. ([kotlin:S1135](https://sonarcloud.io/organizations/github-apollo-rsps/rules#rule_key=kotlin%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=apollo%3Aorg.apollo&pullRequest=338&issues=AWVzgXxBXKENQn1Ssjbl&open=AWVzgXxBXKENQn1Ssjbl)

[ISSUE_KEY]: AWVzgXxBXKENQn1Ssjbl

garyttierney reviewed (commented):
<removed>

garyttierney committed:
Use new SonarCloud organization

Subby committed:
Add command to teleport to another player

Closes #410

garyttierney committed:
Add herblore unit tests and clean-up plugin code

garyttierney committed:
Add herblore tests

garyttierney committed:
Remove apollo-plugin Gradle plugin

Replaces the old Gradle plugin with plain Gradle buildscripts.  Also cleans-up
the current Gradle extensions and re-adds detekt, along with JUnit 5 support to
all modules.

garyttierney committed:
Set input directory in detekt task

garyttierney committed:
Create detekt profiles for each project

garyttierney committed:
Fix merging of JaCoCo reports

Major- committed:
Remove inline modifier from string asserts

rmcmk committed:
Remove duplicate GameObject rotated checks

Major- committed:
Repackage player actions plugin

Major- committed:
Rename walk-to to pathing

Major- committed:
Repackage following plugin

garyttierney committed:
Add runecrafting action unit tests

garyttierney committed:
Add tests for most command plugins

garyttierney committed:
Re-add HTML and XML code coverage reports

garyttierney committed:
Remove broken imports from command tests

garyttierney committed:
Fix Detekt report path

garyttierney committed:
Add additional command test coverage

garyttierney committed:
Fix typo in skill command tests

garyttierney committed:
Add tests for remaining commands

garyttierney committed:
Fix mockk error in matcher

garyttierney committed:
Fix kotlin code style issues

garyttierney committed:
Add library containing detekt rules for apollo plugins

rmcmk committed:
Allow Netty to automatically maintain our internal usages of ByteBuf

Commit:
Housekeeping

Commit:
Let ::pos return other players positions (#422)

Rework ::pos to allow a user to return another player's position

Major- committed:
Merge pull request #421 from KeepBotting/kotlin-experiments

Fix data path and resolution of plugin dependencies

Major- committed:
Merge pull request #420 from ryleykimmel/netty-resource-leak-fix

Allow Netty to automatically maintain our internal usages of ByteBuf

Major- committed:
Merge pull request #414 from ryleykimmel/enhancement/orientation-duplicate

Remove duplicate GameObject rotated checks

Commit:
Add MagicOnMob MessageHandler

Major- committed:
Update kotlin and coroutines dependencies

Major- committed:
Refactor shops plugin

Major- committed:
Update detekt

This temporarily disables the custom detekt rules, as they break the
build in hard-to-debug ways.

Major- committed:
Merge branch 'Tomm0017-magic_on_mob' into kotlin-experiments

Merge #426

garyttierney committed:
Update test deps for Java 11

Prevent bytecode generation related test failures in tests that rely on
PowerMock by updating to PowerMock 2.

garyttierney committed:
Add Azure DevOps pipeline configuration

garyttierney committed:
Run builds in parallel on Azure CI

garyttierney committed:
Enable Gradle build caching in CI

garyttierney committed:
Publish test results to CI

garyttierney committed:
Publish code coverage from Azure CI

garyttierney committed:
Remove AppVeyor and Travis-CI configuration

garyttierney committed:
Publish SonarQube analysis from Azure CI

garyttierney committed:
Merge pull request #427 from apollo-rsps/bugfix/jdk-11

Update test deps for Java 11

dependabot-preview[bot] committed:
Bump org.jetbrains.kotlin.jvm from 1.3.40 to 1.3.41

Bumps org.jetbrains.kotlin.jvm from 1.3.40 to 1.3.41.

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
Bump org.jmailen.kotlinter from 1.16.0 to 1.26.0

Bumps org.jmailen.kotlinter from 1.16.0 to 1.26.0.

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

garyttierney committed:
Fix location of detekt report

garyttierney committed:
Merge pull request #433 from apollo-rsps/bugfix/detekt

Fix location of detekt report

garyttierney committed:
Merge pull request #430 from apollo-rsps/dependabot/gradle/org.jmailen.kotlinter-1.26.0

Bump org.jmailen.kotlinter from 1.16.0 to 1.26.0

garyttierney committed:
Merge pull request #431 from apollo-rsps/dependabot/gradle/org.jetbrains.kotlin.jvm-1.3.41

Bump org.jetbrains.kotlin.jvm from 1.3.40 to 1.3.41

dependabot-preview[bot] committed:
Bump org.sonarqube from 2.6.2 to 2.7.1

Bumps org.sonarqube from 2.6.2 to 2.7.1.

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

garyttierney committed:
Merge pull request #429 from apollo-rsps/dependabot/gradle/org.sonarqube-2.7.1

Bump org.sonarqube from 2.6.2 to 2.7.1

Major- committed:
Implement new listener dsl for plugins

Major- committed:
Reorganise kotlin message handlers

Major- committed:
Add ButtonClick message handler

Major- committed:
Create raw handlers in plugin listener overloads

Major- committed:
Update slack invite link

Major- committed:
Add discord link

GregHib committed:
Fix NPC vertical movement

garyttierney committed:
Merge pull request #495 from GregHib/patch-1

Fix NPC vertical movement
